### PR TITLE
media-gfx/blender: add 4.0.2-r1, cleanup

### PR DESCRIPTION
**Pull request by negril**
hopefully fixed osl build
re-added hip flag in 4.0.2-r1
hide test code in release versions

Bug: https://bugs.gentoo.org/693200
Closes: https://bugs.gentoo.org/925534
Closes: https://bugs.gentoo.org/927835
Closes: https://bugs.gentoo.org/927715
Closes: https://bugs.gentoo.org/927281
Closes: https://bugs.gentoo.org/925534

### CODE REVIEW

**Comment by DarkDefender** [comment]
In the past, one of the biggest issues was when osl and mesa linked to different llvm versions.
Do you think it would be possible to check for that somehow in the ebuild?

To get the live ebuild to work reliably, we need to wait for this to land as well: https://github.com/gentoo/gentoo/pull/35911

**Comment by negril** [comment]
mesa uses llvm-r1.eclass as well. I think it should be possible to make that an RDEPEND.

**Comment by Jamim** [comment]
_Hello @negril,_

There are 2 broken links in the commit message and in the PR description.
```diff
-Closes: https://bugs.gentoo.org927835
-Closes: https://bugs.gentoo.org927715
+Closes: https://bugs.gentoo.org/927835
+Closes: https://bugs.gentoo.org/927715
```
Would you mind fixing them?

_Thanks!_

**Comment by negril** [comment]
I'll rebase once https://github.com/gentoo/gentoo/pull/35969 is merged. So ci clears.

File: media-gfx/blender/blender-4.0.2-r1.ebuild [hunk 1]
```diff
@@ -287,7 +306,6 @@ src_configure() {
 		-DWITH_INPUT_NDOF=$(usex ndof)
 		-DWITH_INTERNATIONAL=$(usex nls)
 		-DWITH_JACK=$(usex jack)
-		-DWITH_LLVM=$(usex osl)
```
Comment (sin-ack) [inline media-gfx/blender/blender-4.0.2-r1.ebuild:290]: If you have LLVM 18 installed, the build will fail in a similar manner to https://bugs.gentoo.org/927281. According to the bundled `FindLLVM.cmake` we should be adding something like this here:

```ebuild
		$(usex osl "-DLLVM_ROOT_DIR=$(get_llvm_prefix)")
```

Comment (negril) [reply media-gfx/blender/blender-4.0.2-r1.ebuild:290]: I have tested this and it worked fine for me. Got any logs that show otherwise?

**Comment by sin-ack** [comment]
I see. As I mentioned above, using an upstream patch allows Blender 4.0.2 to build using the new OSL ebuild.

File: media-gfx/blender/blender-4.0.2-r1.ebuild [hunk 2]
```diff
@@ -484,7 +522,7 @@ pkg_postrm() {
 
 	ewarn ""
 	ewarn "You may want to remove the following directory."
-	ewarn "~/.config/${PN}/${SLOT}/cache/"
+	ewarn "\~/.config/${PN}/${SLOT}/cache/"
```
Comment (sin-ack) [inline media-gfx/blender/blender-4.0.2-r1.ebuild:525]: This shows up as a literal backslash in the post-merge message. Is there a specific reason to escape it?

Comment (negril) [reply media-gfx/blender/blender-4.0.2-r1.ebuild:525]: It silences the [shellcheck warning](https://www.shellcheck.net/wiki/SC2088) and makes it clear that it's no supposed to be expanded.

**Comment by negril** [comment]
I have blender-4.1.0 ready to PR when the dependencies are in, that includes the patch. Do you feel like it's worth it to backport the patch to get rid of the llvm-15 dependency?

**Comment by sin-ack** [comment]
Probably not if we're gonna get that in anyway.

File: media-gfx/blender/blender-4.0.2-r1.ebuild [hunk 3]
```diff
@@ -88,14 +96,14 @@ RDEPEND="${PYTHON_DEPS}
 		>=dev-libs/imath-3.1.4-r2:=
 		>=media-libs/openexr-3:0=
 	)
-	openpgl? ( >=media-libs/openpgl-0.5.0 )
+	openpgl? ( media-libs/openpgl:0/0.5= )
```
[the remark this answers is not in the document]
Comment (negril) [reply media-gfx/blender/blender-4.0.2-r1.ebuild:99]: This depends on https://github.com/gentoo/gentoo/pull/35969 so wait for that please.

File: media-gfx/blender/blender-9999.ebuild [hunk 4]
```diff
@@ -182,10 +196,13 @@ src_unpack() {
 
 		git-r3_fetch "${ADDONS_EGIT_REPO_URI}"
 		git-r3_checkout "${ADDONS_EGIT_REPO_URI}" "${S}/scripts/addons"
-		# TODO
-		#if use test; then
 
-		#fi
+		if use test; then
```
Comment (DarkDefender) [inline media-gfx/blender/blender-9999.ebuild:200]: Why did you add these svn commands back?

I think they will fail regardless because we do not inherit the svn eclass anymore.

Because upstream migrated to git-lfs, we should not use subversion anymore as those repos are not updated anymore.

To get the tests to work I think we only need a:
```
 if use test; then
 EGIT_SUBMODULES=( '*' '-lib/*' )
 fi
 ```
 at the `inherit git-r3` to get the test repo checkedout.

Comment (negril) [reply media-gfx/blender/blender-9999.ebuild:200]: I knew there was a reason that was commented out. Will fix in a bit.

Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild:200]: Note that this will only work if you have the changes that was merged into the main repo 2 days ago:
https://github.com/gentoo/gentoo/pull/35911

Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild]: Right, I we still have to download the test files and host them ourselves if we download them as a tar.
In the past this has proven quite difficult as I don't have a place to host these myself.

Perhaps we could simply use git+git-lfs to fetch for test files for non live ebuilds as well? :thinking: 

I have additional remarks that I think might be better by me doing a PR to your branch. I'll try to have something later today.

**Comment by DarkDefender** [comment]
BTW, did you add any checks to make sure that the llvm version for mesa and osl is the same?
Blender will segfault on start up if you have a GPU that uses mesa and you have a llvm mismatch between these to libraries.

File: media-gfx/blender/blender-9999.ebuild [hunk 5]
```diff
@@ -381,10 +395,22 @@ src_test() {
...
+	if use cuda; then
+		cuda_add_sandbox -w
+		addwrite "/dev/dri/renderD128"
+		addwrite "/dev/char/"
+	fi
+
+	if use X; then
+		xdg_environment_reset
+		LIBGL_ALWAYS_SOFTWARE=1 \
+		virtx cmake_src_test
```
Comment (DarkDefender) [inline media-gfx/blender/blender-9999.ebuild:407]: I'm guessing you added this because you had issues with GL initialization?
It is a bit weird to me though as Blender should now always used EGL and that should work in a headless environment without a X or wayland display.

Where was it failing for you?

**Comment by negril** [comment]
`xdg_environment_reset` is due to xdg related paths missing.
`LIBGL_ALWAYS_SOFTWARE=1` avoids the zink related messages. Doesn't fix any failures though.

This is the diff from with to without `virtx`.
```diff
< 77% tests passed, 105 tests failed out of 456
---
> 76% tests passed, 108 tests failed out of 456
1575354c97010
< Total Test time (real) = 1144.53 sec
---
> Total Test time (real) = 1021.25 sec
1575357c97013
<        44 - draw (Failed)
---
>        44 - draw (SEGFAULT)
1575461a97118,97120
>       256 - compositor_vector_realtime (Failed)
>       257 - compositor_multiple_node_setups_realtime (Failed)
>       258 - compositor_distort_realtime (Failed)
```
I'm fine with dropping the virtx stuff it if you want me to. Seeing how even without the sandbox 71 tests fail for me, we'll have to investigate that anyway.

**Comment by DarkDefender** [comment]
> I'm fine with dropping the virtx stuff it if you want me to. Seeing how even without the sandbox 71 tests fail for me, we'll have to investigate that anyway.

I did manage to get the whole test suite passing around January this year. So I would say to drop `virtx` as it shouldn't be needed.
Of course there is a lot of variables that has changed since January, so I'm not too surprised if there are failures at this point.

For the other changes to the ebuild, I would suggest this:
```diff
diff --git a/media-gfx/blender/blender-9999.ebuild b/media-gfx/blender/blender-9999.ebuild
index 5af54fb20e75..12e5b07c5278 100644
--- a/media-gfx/blender/blender-9999.ebuild
+++ b/media-gfx/blender/blender-9999.ebuild
@@ -17,7 +17,6 @@ if [[ ${PV} = *9999* ]] ; then
        EGIT_LFS="yes"
        inherit git-r3
        EGIT_REPO_URI="https://projects.blender.org/blender/blender.git"
-       EGIT_SUBMODULES=( '*' '-lib/*' )
        ADDONS_EGIT_REPO_URI="https://projects.blender.org/blender/blender-addons.git"
        RESTRICT="!test? ( test )"
 else
@@ -188,7 +187,13 @@ pkg_setup() {

 src_unpack() {
        if [[ ${PV} = *9999* ]] ; then
-               git-r3_src_unpack
+               if use test ; then
+                       # Ensure that we clone down the test related submodules
+                       local EGIT_SUBMODULES=( '*' '-lib/*' )
+                       git-r3_src_unpack
+               else
+                       git-r3_src_unpack
+               fi

                git-r3_fetch "${ADDONS_EGIT_REPO_URI}"
                git-r3_checkout "${ADDONS_EGIT_REPO_URI}" "${S}/scripts/addons"
@@ -486,11 +491,11 @@ pkg_postinst() {
                ewarn ""
        fi

-       if ! use python_single_target_python3_10; then
+       if ! use python_single_target_python3_11; then
                elog "You are building Blender with a newer python version than"
                elog "supported by this version upstream."
                elog "If you experience breakages with e.g. plugins, please switch to"
-               elog "python_single_target_python3_10 instead."
+               elog "python_single_target_python3_11 instead."
                elog "Bug: https://bugs.gentoo.org/737388"
                elog
        fi
 ```
 I introduced the `local EGIT_SUBMODULES=( '*' '-lib/*' )` so that we don't clone down the test submodule unless we need it.
 
 For the python elog change, Blender is using python 3.11 in the official builds since 4.1.
 
Sadly my main computer blew up during the weekend, so I can't really test things out. Of course don't wait for me to get a new one.

**Comment by eli-schwartz** [comment]
Build fails?

```
FAILED: source/blender/imbuf/intern/oiio/CMakeFiles/bf_imbuf_openimageio.dir/openimageio_support.cc.o 
/usr/bin/x86_64-pc-linux-gnu-g++ -DHAVE_EXECINFO_H -DNDEBUG -DWITH_ASSERT_ABORT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -D__LITTLE_ENDIAN__ -D__MMX__ -D__SSE2__ -D__SSE__ -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/blenkernel -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/intern/utfconv -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/blenlib -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/makesdna -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0_build/source/blender/makesdna/intern -I/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/intern/guardedalloc -isystem /usr/include/OpenEXR -isystem /usr/include/Imath  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNDEBUG -Wuninitialized -Wredundant-decls -Wall -Wno-invalid-offsetof -Wno-sign-compare -Wlogical-op -Winit-self -Wmissing-include-dirs -Wno-div-by-zero -Wtype-limits -Werror=return-type -Wno-char-subscripts -Wno-unknown-pragmas -Wpointer-arith -Wunused-parameter -Wwrite-strings -Wundef -Wcomma-subscript -Wformat-signedness -Wrestrict -Wno-suggest-override -Wuninitialized -Wno-stringop-overread -Wno-stringop-overflow -Wimplicit-fallthrough=5 -Wundef -Wmissing-declarations -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -fopenmp -msse -pipe -fPIC -funsigned-char -fno-strict-aliasing -ffp-contract=off -msse2 -fmacro-prefix-map="/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/"="" -fmacro-prefix-map="/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0_build/"="" -Wno-maybe-uninitialized  -std=c++17 -MD -MT source/blender/imbuf/intern/oiio/CMakeFiles/bf_imbuf_openimageio.dir/openimageio_support.cc.o -MF source/blender/imbuf/intern/oiio/CMakeFiles/bf_imbuf_openimageio.dir/openimageio_support.cc.o.d -o source/blender/imbuf/intern/oiio/CMakeFiles/bf_imbuf_openimageio.dir/openimageio_support.cc.o -c /var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc
/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc: In function ‘std::unique_ptr<OpenImageIO_v2_4::ImageInput> blender::imbuf::get_oiio_reader(const char*, const OpenImageIO_v2_4::ImageSpec&, OpenImageIO_v2_4::Filesystem::IOMemReader&, OpenImageIO_v2_4::ImageSpec&)’:
/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc:253:29: error: no matching function for call to ‘OpenImageIO_v2_4::ImageInput::valid_file(OpenImageIO_v2_4::Filesystem::IOMemReader*)’
  253 |   if (!(in && in->valid_file(&mem_reader))) {
      |               ~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.hh:15,
                 from /var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc:5:
/usr/include/OpenImageIO/imageio.h:1017:18: note: candidate: ‘virtual bool OpenImageIO_v2_4::ImageInput::valid_file(const std::string&) const’
 1017 |     virtual bool valid_file (const std::string& filename) const;
      |                  ^~~~~~~~~~
/usr/include/OpenImageIO/imageio.h:1017:49: note:   no known conversion for argument 1 from ‘OpenImageIO_v2_4::Filesystem::IOMemReader*’ to ‘const std::string&’ {aka ‘const std::__cxx11::basic_string<char>&’}
 1017 |     virtual bool valid_file (const std::string& filename) const;
      |                              ~~~~~~~~~~~~~~~~~~~^~~~~~~~
/usr/include/OpenImageIO/imageio.h:1020:10: note: candidate: ‘bool OpenImageIO_v2_4::ImageInput::valid_file(const std::wstring&) const’
 1020 |     bool valid_file (const std::wstring& filename) const {
      |          ^~~~~~~~~~
/usr/include/OpenImageIO/imageio.h:1020:42: note:   no known conversion for argument 1 from ‘OpenImageIO_v2_4::Filesystem::IOMemReader*’ to ‘const std::wstring&’ {aka ‘const std::__cxx11::basic_string<wchar_t>&’}
 1020 |     bool valid_file (const std::wstring& filename) const {
      |                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc: In function ‘bool blender::imbuf::imb_oiio_check(const uchar*, size_t, const char*)’:
/var/tmp/portage/media-gfx/blender-4.1.0/work/blender-4.1.0/source/blender/imbuf/intern/oiio/openimageio_support.cc:274:30: error: no matching function for call to ‘OpenImageIO_v2_4::ImageInput::valid_file(OpenImageIO_v2_4::Filesystem::IOMemReader*)’
  274 |   return in && in->valid_file(&mem_reader);
      |                ~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/include/OpenImageIO/imageio.h:1017:18: note: candidate: ‘virtual bool OpenImageIO_v2_4::ImageInput::valid_file(const std::string&) const’
 1017 |     virtual bool valid_file (const std::string& filename) const;
      |                  ^~~~~~~~~~
/usr/include/OpenImageIO/imageio.h:1017:49: note:   no known conversion for argument 1 from ‘OpenImageIO_v2_4::Filesystem::IOMemReader*’ to ‘const std::string&’ {aka ‘const std::__cxx11::basic_string<char>&’}
 1017 |     virtual bool valid_file (const std::string& filename) const;
      |                              ~~~~~~~~~~~~~~~~~~~^~~~~~~~
/usr/include/OpenImageIO/imageio.h:1020:10: note: candidate: ‘bool OpenImageIO_v2_4::ImageInput::valid_file(const std::wstring&) const’
 1020 |     bool valid_file (const std::wstring& filename) const {
      |          ^~~~~~~~~~
/usr/include/OpenImageIO/imageio.h:1020:42: note:   no known conversion for argument 1 from ‘OpenImageIO_v2_4::Filesystem::IOMemReader*’ to ‘const std::wstring&’ {aka ‘const std::__cxx11::basic_string<wchar_t>&’}
 1020 |     bool valid_file (const std::wstring& filename) const {
      |                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```

openimageio version: media-libs/openimageio-2.4.6.0-r1

File: media-gfx/blender/blender-9999.ebuild (hunk 4 above)
Comment (negril) [reply media-gfx/blender/blender-9999.ebuild]: https://github.com/gentoo/gentoo/blob/master/eclass/git-r3.eclass#L36 this makes it impossible to use git-r3 eclass to do that. We could call git ourselves. But I'm not really sure if that the right thing to do. I think filing a bug with blender to enable the setting to make archives including the actual files might be better.

Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild]: gitea doesn't support packaging the LFS files into the repo .zip ATM.
I've also talked with the Blender admins in charge of this (I'm a Blender Employee as well), and sadly they are not too convinced that we should create zip files with the test files for the releases.
They are mainly concerned about the additional storage and bandwidth requirements as we currently are struggling a bit with that.

So I guess we would have to still host the test files ourselves like sam tried to do for a few releases. :(

Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild]: Or hmm...
@mgorny how do you feel about being able to not have the "live" tag for git checkouts that is using tags?
This way we would be able to download releases from repos with lfs files or submodules without having to package it ourselves.

**Comment by DarkDefender** [comment]
Just an other heads up. Libdecor has been added to the portage tree now.
https://bugs.gentoo.org/927186
It is needed in Gnome to draw the title bars.
https://bugs.gentoo.org/927931

Perhaps we can add this with a `gnome` useflag? Or perhaps a `libdecor` useflag is better?

File: media-gfx/blender/blender-9999.ebuild (hunk 4 above)
Comment (eli-schwartz) [reply media-gfx/blender/blender-9999.ebuild]: I suspect the answer is going to be "it is inconvenient for a reason, because people aren't supposed to use it just for the sake of convenience".

The point of the "live" property is more than just "we assume the exact commit version is unknown".

**Comment by negril** [comment]
I'd just add it to the `wayland` flag?

**Comment by DarkDefender** [comment]
Only the gnome wayland desktop needs it. The other wayland compositors does not need libdecor IIRC.

**Comment by negril** [comment]
I left 3.3.17 and 3.6.10 out for now. I need to backport more changes from 4.1.0.

File: media-gfx/blender/blender-9999.ebuild [hunk 6]
```diff
@@ -4,8 +4,11 @@
 EAPI=8
 
 PYTHON_COMPAT=( python3_{11..12} )
+# matches media-libs/osl
+LLVM_COMPAT=( {15..17} )
+VIRTUALX_REQUIRED=manual
```
Comment (DarkDefender) [inline media-gfx/blender/blender-9999.ebuild:9]: Is this variable needed now that you removed the explicit calls to `virtx`?

**Comment by DarkDefender** [comment]
> I left 3.3.17 and 3.6.10 out for now. I need to backport more changes from 4.1.0.

4.1.1 is schedule for release in two days (April 16th: https://projects.blender.org/blender/blender/issues/119631)
At this point perhaps it is best to wait a few days and skip 4.1.0?

**Comment by khumarahn** [comment]
> 4.1.1 is scheduled for release in two days (April 16th: https://projects.blender.org/blender/blender/issues/119631) At this point perhaps it is best to wait a few days and skip 4.1.0?

This PR fixes some bugs (like blender not compiling), it would be really nice to get it merged soon though

File: media-gfx/blender/blender-9999.ebuild (hunk 6 above)
Comment (negril) [reply media-gfx/blender/blender-9999.ebuild]: Fixed.

**Comment by negril** [comment]
I move 4.1.0 to a different branch https://github.com/negril/gentoo/tree/media-gfx/blender/4.1.0 so we can get the bugfixes in and get blender working agin and then PR 4.1.1 in two days.

**Comment by DarkDefender** [comment]
I've tested the live ebuild to look at the test faliures. The majority of them seems to be some EGL initialization problem.I've seen this happen with offical builds as well. I think this is a Blender issue, so I'll try to look into it a bit more with some of the other upstream devs.

This should be good to land though as I didn't see any test failures that would be a show stopper.

**Comment by negril** [comment]
I don't think it's a blender issue. I'v seen EGL stuff fail with other ebuilds. Back then I tested all the mesa versions in tree, anything >23.1.9 failed. Maybe check that against the official CI?

LIBGL_ALWAYS_SOFTWARE=true might help afair.
